test/merger/test_single_qubit_gates_merger.py

```diff
@@ -0,0 +1,283 @@
+import math
+
+import pytest
+
+from opensquirrel import Circuit, CircuitBuilder
+from opensquirrel.default_instructions import Ry, Rz
+from opensquirrel.ir import BlochSphereRotation, Float, Qubit
+from opensquirrel.passes.merger import SingleQubitGatesMerger
+from opensquirrel.passes.merger.general_merger import rearrange_barriers
+from test.ir_equality_test_base import modify_circuit_and_check
+
+
+@pytest.fixture(name="merger")
+def merger_fixture() -> SingleQubitGatesMerger:
+    return SingleQubitGatesMerger()
+
+
+def test_single_gate(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(1)
+    builder1.Ry(0, 1.2345)
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(1)
+    builder2.Ry(0, 1.2345)
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+    assert circuit.ir.statements == [Ry(0, 1.2345)]
+
+
+def test_two_hadamards(merger: SingleQubitGatesMerger) -> None:
+    builder = CircuitBuilder(4)
+    builder.H(2)
+    builder.H(2)
+    circuit = builder.to_circuit()
+
+    expected_circuit = CircuitBuilder(4).to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+
+def test_two_hadamards_different_qubits(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(4)
+    builder1.H(0)
+    builder1.H(2)
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(4)
+    builder2.H(0)
+    builder2.H(2)
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+
+def test_merge_different_qubits(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(4)
+    builder1.Ry(0, math.pi / 2)
+    builder1.Rx(0, math.pi)
+    builder1.Rz(1, 1.2345)
+    builder1.Ry(2, 1)
+    builder1.Ry(2, 3.234)
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(4)
+    builder2.ir.add_gate(BlochSphereRotation(0, axis=(1, 0, 1), angle=math.pi))  # this is Hadamard with 0 phase
+    builder2.Rz(1, 1.2345)
+    builder2.Ry(2, 4.234)
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+    assert isinstance(circuit.ir.statements[0], BlochSphereRotation)
+    assert circuit.ir.statements[0].is_anonymous  # When fusion happens, the resulting gate is anonymous.
+
+    assert isinstance(circuit.ir.statements[1], BlochSphereRotation)
+    assert circuit.ir.statements[1].generator == Rz  # Otherwise it keeps the same generator and arguments.
+    assert circuit.ir.statements[1].arguments == (Qubit(1), Float(1.2345))
+
+    assert isinstance(circuit.ir.statements[2], BlochSphereRotation)
+    assert circuit.ir.statements[2].is_anonymous
```

Comment:
I agree with leaving the `compose_bloch_sphere_rotations` fix for another PR, but I still think the asserts can be removed (the code below is untested):

```suggestion
    builder2.ir.add_gate(BlochSphereRotation(2, axis=(0, 1, 0), angle=4.234))
    expected_circuit = builder2.to_circuit()
```